app/ptokens/models.py

```diff
@@ -0,0 +1,366 @@
+                                                       redemption_state__in=[
+                                                           'request', 'accepted', 'waiting_complete'
+                                                       ]).aggregate(locked=Sum('total'))['locked'] or 0
+        available_to_redeem = current_hodling - locked_amount
+
+        return available_to_redeem
+
+    def update_token_status(self):
+        if PTOKEN_ABI and self.token_address:
+            web3 = get_web3(self.network, sockets=True)
+            contract = web3.eth.contract(Web3.toChecksumAddress(self.token_address), abi=PTOKEN_ABI)
+            decimals = 10 ** contract.functions.decimals().call()
+            self.total_minted = contract.functions.totalSupply().call() // decimals
+            self.value = contract.functions.price().call() // decimals
+
+            if self.tx_status == 'success' and self.txid:
+                latest = web3.eth.blockNumber
+                tx = web3.eth.getTransaction(self.txid)
+                if self.last_block == 0:
+                    self.last_block = tx['blockNumber']
+
+                redeem_filter = contract.events.Redeemed.createFilter(fromBlock=self.last_block, toBlock=latest)
+                purchase_filter = contract.events.Purchased.createFilter(fromBlock=self.last_block, toBlock=latest)
+                redeemed = 0
+                purchased = 0
+                for redeem in redeem_filter.get_all_entries():
+                    redeemed += redeem['args']['amountRedeemed']
+
+                for purchase in purchase_filter.get_all_entries():
+                    purchased += purchase['args']['amountReceived']
+
+                redeemed = redeemed // decimals
+                purchased = purchased // decimals
+
+                self.total_purchased = purchased
+                self.total_redeemed = redeemed
+                self.total_available = self.total_minted - (self.total_purchased - self.total_redeemed)
+                self.redemptions = len(redeem_filter.get_all_entries())
+                self.purchases = len(purchase_filter.get_all_entries())
+
+                print(f'REDEEMED: {self.total_redeemed}')
+                print(f'PURCHASED: {self.total_purchased}')
+                print(f'AVAILABLE: {self.total_available}')
+            self.save()
+
+    def update_user_balance(self, holder_profile, holder_address):
+        if PTOKEN_ABI and self.token_address:
+            web3 = get_web3(self.network)
+            contract = web3.eth.contract(Web3.toChecksumAddress(self.token_address), abi=PTOKEN_ABI)
+            decimals = 10 ** contract.functions.decimals().call()
+            balance = contract.functions.balanceOf(holder_address).call() // decimals
+            self.cached_balances[str(holder_profile.id)] = balance
+            self.save()
+
+    def update_tx_status(self):
+        # Get transaction status
+        from dashboard.utils import get_tx_status
```

Comment:
Can we move these imports?

Reply:
Resolved in 34f55140b08c1f940fbde01f5754255cd44b4d70